src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraph.java

```diff
@@ -0,0 +1,464 @@
+        for (final String child : cgs.getAllDescendentsMap().get(key)) {
+            CacheNode oldChild = nodesMap.get(child);
+            /* all of the desendents are computable nodes and can be safely upcasted */
+            updatedNodesMap.put(child, ((ComputableCacheNode)oldChild).duplicateWithUpdatedCacheStatus(false));
+        }
+        return updatedNodesMap;
+    }
+
+    /**
+     * Returns a reference to the value of a given node
+     *
+     * Note: this function is purposefully meant to be light:
+     *
+     * (1) it does not update out-of-date cacheable computable nodes, and
+     * (2) it does not evaluate non-cacheable computable nodes.
+     *
+     * @param nodeKey the key of the node
+     * @return value of the node
+     * @throws IllegalStateException if a cached node is out of date, or if the value of a primitive node is
+     *                               not initialized
+     * @throws IllegalArgumentException if the node does not exist
+     */
+    public Duplicable getValueDirect(@Nonnull final String nodeKey)
+            throws IllegalStateException, IllegalArgumentException {
+        assertNodeExists(Utils.nonNull(nodeKey, "The key of the node can not be null."));
+        return nodesMap.get(nodeKey).getValue(EMPTY_MAP);
+    }
+
+    /**
+     *
+     * @param nodeKey key of the node
+     * @return value of the node
+     * @throws IllegalArgumentException if the node does not exist
+     */
+    public Duplicable getValueWithRequiredEvaluations(@Nonnull final String nodeKey)
+            throws IllegalStateException, IllegalArgumentException {
+        assertNodeExists(Utils.nonNull(nodeKey, "The key of the node can not be null."));
+        return accumulateEvaluationsForDepthSortedNodeList(cgs.getEvalStrategyByEvaluatedNode(nodeKey), true).get(nodeKey);
+    }
+
+    /**
+     * This method calculates and accumulates the values of a list of depth-sorted nodes (from low depth to high depth).
+     * Compute-on-demand nodes are always evaluated. Caching nodes are only evaluated if {@code evalOnOutOfDateCachingNodes}
+     * is true.
+     *
+     * @param depthSortedNodes depth-sorted list of nodes
+     * @param evalOutdatedCachingNodes evaluate
+     * @return
+     */
+    private Map<String, Duplicable> accumulateEvaluationsForDepthSortedNodeList(@Nonnull final List<String> depthSortedNodes,
+                                                                                final boolean evalOutdatedCachingNodes) {
+        final Map<String, Duplicable> accumulatedValues = new HashMap<>();
+        for (final String nodeKey : depthSortedNodes) {
+            if (nodesMap.get(nodeKey).isPrimitive()) {
+                Duplicable value;
+                try {
+                    value = nodesMap.get(nodeKey).getValue(accumulatedValues);
+                } catch (IllegalStateException e) {
+                    value = null;
```

Comment:
The fact that the `catch` block doesn't `throw` makes me uncomfortable because it means the code in the `try` block might throw an error that's not really an error.